polkadot/node/core/prospective-parachains/src/fragment_tree.rs

```diff
@@ -753,53 +753,127 @@ impl FragmentTree {
+	/// If there is no chain of size `count` that matches the criteria, this will return the largest
+	/// chain it could find with the criteria.
+	/// If there are no candidates meeting those criteria, returns an empty `Vec`.
+	/// Cycles are accepted, see module docs for the `Cycles` section.
 	///
 	/// The intention of the `required_path` is to allow queries on the basis of
 	/// one or more candidates which were previously pending availability becoming
 	/// available and opening up more room on the core.
-	pub(crate) fn select_child(
+	pub(crate) fn select_children(
 		&self,
 		required_path: &[CandidateHash],
+		count: u32,
 		pred: impl Fn(&CandidateHash) -> bool,
-	) -> Option<CandidateHash> {
+	) -> Vec<CandidateHash> {
 		let base_node = {
 			// traverse the required path.
 			let mut node = NodePointer::Root;
 			for required_step in required_path {
-				node = self.node_candidate_child(node, &required_step)?;
+				if let Some(next_node) = self.node_candidate_child(node, &required_step) {
+					node = next_node;
+				} else {
+					return vec![]
+				};
 			}
 
 			node
 		};
 
-		// TODO [now]: taking the first selection might introduce bias
+		// TODO: taking the first best selection might introduce bias
 		// or become gameable.
 		//
 		// For plausibly unique parachains, this shouldn't matter much.
 		// figure out alternative selection criteria?
-		match base_node {
+		self.select_children_inner(base_node, count, count, &pred, &mut vec![])
+	}
+
+	// Try finding a candidate chain starting from `base_node` of length `expected_count`.
+	// If not possible, return the longest one we could find.
+	// Does a depth-first search, since we're optimistic that there won't be more than one such
+	// chains (parachains shouldn't usually have forks). So in the usual case, this will conclude
+	// in `O(expected_count)`.
+	// Cycles are accepted, but this doesn't allow for infinite execution time, because the maximum
+	// depth we'll reach is `expected_count`.
+	//
+	// Worst case performance is `O(num_forks ^ expected_count)`.
+	// Although an exponential function, this is actually a constant that can only be altered via
+	// sudo/governance, because:
+	// 1. `num_forks` at a given level is at most `max_candidate_depth * max_validators_per_core`
+	//    (because each validator in the assigned group can second `max_candidate_depth`
+	//    candidates). The prospective-parachains subsystem assumes that the number of para forks is
+	//    limited by collator-protocol and backing subsystems. In practice, this is a constant which
+	//    can only be altered by sudo or governance.
+	// 2. `expected_count` is equal to the number of cores a para is scheduled on (in an elastic
+	//    scaling scenario). For non-elastic-scaling, this is just 1. In practice, this should be a
+	//    small number (1-3), capped by the total number of available cores (a constant alterable
+	//    only via governance/sudo).
```

Comment:
So with realistic numbers (current and expected/exaggerated future numbers): How bad does it get? What would be limits to those parameters?

Reply:
Elastic scaling will eventually enable chains to be bounded by the rate of a single node's block production.

In Cumulus currently, block authorship is typically slower than block verification (in validators) due to the need to read from disk. Most of the time is spent in accessing the trie for storage, not in code execution. Access patterns for the trie are also very inefficient. So it's not feasible to use more than 3 cores until that bottleneck is cleared.

If this bottleneck is alleviated, then I could see parachains using many more cores.

Reply:
thinking about this a bit more, I don't think this traversal is any different in terms of worst-case complexity from the routine used for populating the fragment tree (which populates the tree breadth-first). Is it?
In an acyclic tree, it'll visit each node once (in a worst case scenario). In a cyclic one, it'd be bounded by the requested count so it's the same complexity.

The complexity is linear in the number of nodes, but the number of nodes is an exponential function (because it's an n-ary tree, at least in theory, of arity `num_forks` and height of `max_candidate_depth`).

If we'll hit a bottleneck, this wouldn't be the first place where that'd surface.

I think we could limit the number of total forks allowed for a parachain (regardless of their level in the tree). So that the total number of branches in the tree would bring down the worst-case complexity massively. WDYT?

Reply:
anyway, I think this is somewhat orthogonal to this PR.
to quote the docs from fragment_tree module:

```
//! The code in this module is not designed for speed or efficiency, but conceptual simplicity.
//! Our assumption is that the amount of candidates and parachains we consider will be reasonably
//! bounded and in practice will not exceed a few thousand at any time. This naive implementation
//! will still perform fairly well under these conditions, despite being somewhat wasteful of
//! memory.
```

we should revisit this assumption maybe, but it doesn't seem like it'll be a problem for the forseeable future

Reply:
`Prospective-parachains` seems to be a good subsystem benchmark target. That should give some clear numbers on CPU usage in worst case scenarios.

Reply:
opened a separate issue to track this security item and added it to the elastic scaling task: https://github.com/paritytech/polkadot-sdk/issues/3219